extensions/flyway/deployment/src/main/java/io/quarkus/flyway/FlywayProcessor.java

```diff
@@ -86,21 +84,59 @@ void build(BuildProducer<AdditionalBeanBuildItem> additionalBeanProducer,
 
     @BuildStep
     @Record(ExecutionTime.RUNTIME_INIT)
-    ServiceStartBuildItem configureRuntimeProperties(FlywayRecorder recorder,
-            FlywayRuntimeConfig flywayRuntimeConfig,
-            BeanContainerBuildItem beanContainer,
+    ServiceStartBuildItem createBeansAndStartActions(FlywayRecorder recorder,
             List<JdbcDataSourceBuildItem> jdbcDataSourceBuildItems,
+            BuildProducer<AdditionalBeanBuildItem> additionalBeans,
+            BuildProducer<SyntheticBeanBuildItem> syntheticBeanBuildItemBuildProducer,
             BuildProducer<JdbcDataSourceSchemaReadyBuildItem> schemaReadyBuildItem) {
-        recorder.doStartActions(flywayRuntimeConfig, beanContainer.getValue());
+
+        // make a FlywayContainerProducer bean
+        additionalBeans.produce(AdditionalBeanBuildItem.builder().addBeanClasses(FlywayContainerProducer.class).setUnremovable()
+                .setDefaultScope(DotNames.SINGLETON).build());
+        // add the @FlywayDataSource class otherwise it won't registered as a qualifier
+        additionalBeans.produce(AdditionalBeanBuildItem.builder().addBeanClass(FlywayDataSource.class).build());
+
+        Collection<String> dataSourceNames = getDataSourceNames(jdbcDataSourceBuildItems);
+
+        for (String dataSourceName : dataSourceNames) {
+            SyntheticBeanBuildItem.ExtendedBeanConfigurator configurator = SyntheticBeanBuildItem
+                    .configure(Flyway.class)
+                    .scope(Dependent.class) // this is what the existing code does, but it doesn't seem reasonable
```

Comment:
I tend to agree we should change that. In a separate PR probably so that we have it in the changelog.

Reply:
I agree